src/py/py_multipatch.cpp

```diff
@@ -671,68 +671,242 @@ void GetBoundaryOrientation(
+                                   const int n_threads) {
+
+  GetBoundaryOrientations(tolerance, n_threads);
+  std::vector<py::array_t<int>> ctr_positions = {};
+  std::vector<int> ctr_index = {};
+  ctr_positions.reserve(orientations_.size());
+  ctr_index.reserve(orientations_.size());
+  const int param_dim = ParaDim();
+  const int dim = Dim();
+
+  const int n_entries_per_line = 4 + 2 * param_dim;
+  int* orientations_ptr = static_cast<int*>(orientations_.request().ptr);
+
+  // get from the face id the position of the control points of the face (step
+  // 1)
+  auto face_id_to_coord_id =
+      [&](const std::vector<int>& cmr,
+          const int& direction,
+          const int& orientation,
+          const int& face_ctr_ptr_id) -> std::vector<int> {
+    std::vector<int> coordinates{};
+    coordinates.reserve(param_dim);
+    int remainder = face_ctr_ptr_id;
+
+    for (int i{}; i < param_dim; i++) {
+      if (direction == i) {
+        coordinates.push_back(orientation > 0 ? cmr[i] - 1 : 0);
+      } else {
+        coordinates.push_back(remainder % cmr[i]);
+        remainder /= cmr[i];
+      }
+    }
+    return coordinates; // control points coordinates -> matrix of ctrs
+  };
+
+  // get the global id from the coordinate id (step 2)
+  auto coord_id_to_glob_id = [&](const std::vector<int>& cmr,
+                                 const std::vector<int>& coordinates,
+                                 const int face_id) -> int {
+    int glob_id{coordinates[param_dim - 1]};
+    for (int i{param_dim - 2}; i > -1; i--) {
+      glob_id *= cmr[i];
+      glob_id += coordinates[i];
+    }
+    return glob_id;
+  };
+
+  auto calculate_squared_euclidean_distance =
+      [&dim](const double* start_values, const double* end_values) -> double {
+    double sum = 0.0;
+    for (int i{}; i < dim; i++) {
+      const double diff = start_values[i] - end_values[i];
+      sum += diff * diff;
+    }
+    return sum;
+  };
+
+  auto get_end_ctp_coordinate = [&](const int face_id,
+                                    const std::vector<int> cmr,
+                                    const int coordinate_id_start_patch,
+                                    const int direction) -> int {
+    if (face_id > 1) {
+      return cmr[direction % 2] - 1;
+    }
+    return coordinate_id_start_patch;
+  };
+
+  for (int i{}; i < (core_patches_.size() - 1); i++) {
+    const int& start_patch_id = orientations_ptr[i * n_entries_per_line + 0];
+    const int& start_face_id = orientations_ptr[i * n_entries_per_line + 1];
+    const int& end_patch_id = orientations_ptr[i * n_entries_per_line + 2];
+    const int& end_face_id = orientations_ptr[i * n_entries_per_line + 3];
+    const int* alignment_ptr = &orientations_ptr[i * n_entries_per_line + 4];
+    const int* orientation_ptr =
+        &orientations_ptr[i * n_entries_per_line + 4 + param_dim];
+
+    // get coordinates pointer from control point
+    const auto start_control_points =
```

Comment:
Here you have to check if the spline is rational or not! Otherwise the function will not work...